File: java/src/main/java/it/giannibombelli/wsc2026/booking/domain/events/BookingResultEvents.java
package it.giannibombelli.wsc2026.booking.domain.events;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.booking.domain.booking.BookingId;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;

import java.util.UUID;


public sealed interface BookingResultEvents extends BookingEvent
    permits BookingResultEvents.BookingConfirmed, BookingResultEvents.BookingRefused, BookingResultEvents.BookingRejected {

    record BookingConfirmed(
        BookingId aggregateId,
        UUID giftCardReference,
        Money amount
    ) implements BookingResultEvents {
        public BookingConfirmed {
            Require.requireArgument(aggregateId, "bookingId");
            Require.requireArgument(giftCardReference, "giftCardReference");
            Require.requireArgument(amount, "amount");
        }
    }

    record BookingRefused(
        BookingId aggregateId,
        UUID giftCardReference,
        Money amount
    ) implements BookingResultEvents {
        public BookingRefused {
            Require.requireArgument(aggregateId, "bookingId");
            Require.requireArgument(giftCardReference, "giftCardReference");
            Require.requireArgument(amount, "amount");
        }
    }

    record BookingRejected(
        BookingId aggregateId,
        UUID giftCardReference,
        Money amount
    ) implements BookingResultEvents {
        public BookingRejected {
            Require.requireArgument(aggregateId, "bookingId");
            Require.requireArgument(giftCardReference, "giftCardReference");
            Require.requireArgument(amount, "amount");
        }
    }
}

File: typescript/src/booking/domain/events/bookingResultEvents.ts
// Eventi di risultato del ciclo di vita di una prenotazione.


import type { Event } from '@/common/domain/model/event.js';
import type { Money } from '@/common/domain/primitive/money.js';
import type { BookingId } from '../booking/bookingId.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

export type BookingConfirmed = Event<'BookingConfirmed', BookingId> & {
  readonly giftCardReference: string;
  readonly amount: Money;
};

export function bookingConfirmed(aggregateId: BookingId, giftCardReference: string, amount: Money): BookingConfirmed {
  requireArgument(aggregateId, 'bookingId');
  requireArgument(giftCardReference, 'giftCardReference');
  requireArgument(amount, 'amount');
  return { kind: 'BookingConfirmed', aggregateId, giftCardReference, amount };
}

export type BookingRefused = Event<'BookingRefused', BookingId> & {
  readonly giftCardReference: string;
  readonly amount: Money;
};

export function bookingRefused(aggregateId: BookingId, giftCardReference: string, amount: Money): BookingRefused {
  requireArgument(aggregateId, 'bookingId');
  requireArgument(giftCardReference, 'giftCardReference');
  requireArgument(amount, 'amount');
  return { kind: 'BookingRefused', aggregateId, giftCardReference, amount };
}

export type BookingRejected = Event<'BookingRejected', BookingId> & {
  readonly giftCardReference: string;
  readonly amount: Money;
};

export function bookingRejected(aggregateId: BookingId, giftCardReference: string, amount: Money): BookingRejected {
  requireArgument(aggregateId, 'bookingId');
  requireArgument(giftCardReference, 'giftCardReference');
  requireArgument(amount, 'amount');
  return { kind: 'BookingRejected', aggregateId, giftCardReference, amount };
}

export type BookingResultEvent = BookingConfirmed | BookingRefused | BookingRejected;
